framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -2591,40 +2603,60 @@ public AnnotatedTypeMirror getDefaultValueAnnotatedType(TypeMirror typeMirror) {
   }
 
   /**
-   * Returns a {@code RequiresQualifier("...")} or {@code EnsuresQualifier("...")} annotation for
-   * the given field. Returns null if none can be created, because the qualifier has
+   * Matches parameter expressions as they appear in {@link EnsuresQualifier} and {@link
+   * RequiresQualifier} annotations, e.g. "#1", "#2", etc.
+   */
+  protected static final Pattern formalParameterPattern = Pattern.compile("^#[0-9]+$");
+
+  /**
+   * Creates a {@code RequiresQualifier("...")} or {@code EnsuresQualifier("...")} annotation for
+   * the given expression. Returns null if none can be created, because the qualifier has
    * elements/arguments, which {@code @RequiresQualifier} and {@code @EnsuresQualifier} do not
-   * support.
+   * support. Also returns null if the expression is invalid when combined with the kind of
+   * annotation: for example, precondition annotations on "this" and parameters ("#1", etc.) are not
+   * supported, because receiver/parameter annotations should be inferred instead.
    *
    * <p>This is of the form {@code @RequiresQualifier(expression="this.elt",
    * qualifier=MyQual.class)} or {@code @EnsuresQualifier(expression="this.elt",
    * qualifier=MyQual.class)} when elt is declared as {@code @A} or {@code @Poly*} and f contains
    * {@code @B} which is a sub-qualifier of {@code @A}.
    *
-   * @param fieldElement a field
+   * <p>Subclasses may override this method to return qualifiers that do have arguments instead of
+   * returning null.
+   *
+   * @param expression the expression to which the qualifier applies
    * @param qualifier the qualifier that must be present
+   * @param declaredType the declared type of the expression, which is used to avoid inferring
+   *     redundant pre- or postcondition annotations
    * @param preOrPost whether to return a precondition or postcondition annotation
-   * @param preconds the precondition annotations for the method; used to suppress redundant
-   *     postconditions; non-null exactly when {@code preOrPost} is {@code AFTER}
+   * @param preconds the list of precondition annotations; used to suppress redundant - *
+   *     postconditions; always null if {@code preOrPost} is {@code BeforeOrAfter.BEFORE}
```

Comment:
Why was this changed from "non-null exactly when" to "always null if"?  The latter is a weaker condition.  Can the list be null when `preOrPost` is `AFTER`?

Another occurrencef of "non-null exactly when" in the code was *not* changed like this one was.